tests/test_tagging.py

```diff
@@ -0,0 +1,301 @@
+        # setup directory to scan for files in the forwarder staging directory
+        # and ensure one file is there
+        ingesterConfig = config["ingester"]
+        forwarder_staging_dir = ingesterConfig["forwarderStagingDirectory"]
+        scanner = DirectoryScanner([forwarder_staging_dir])
+        files = scanner.getAllFiles()
+        self.assertEqual(len(files), 1)
+
+        # create the file ingester, get all tasks associated with it, and
+        # create the tasks
+        ingester = FileIngester(ingesterConfig)
+        butler_tasks = ingester.getButlerTasks()
+
+        task_list = []
+        for butler_task in butler_tasks:
+            task = asyncio.create_task(butler_task())
+            task_list.append(task)
+
+        # check to see that the file is there before ingestion
+        self.assertTrue(os.path.exists(self.destFile))
+
+        # trigger the ingester by sending it a "message"
+        msg = {}
+        msg['CAMERA'] = "COMCAM"
+        msg['OBSID'] = "CC_C_20190530_000001"
+        msg['FILENAME'] = self.destFile
+        msg['ARCHIVER'] = "CCArchiver"
+        await ingester.ingest(msg)
+
+        # make sure staging area is now empty
+        files = scanner.getAllFiles()
+        self.assertEqual(len(files), 0)
+
+        # Check to see that the file was ingested.
+        # Recall that files start in teh forwarder staging area, and are
+        # moved to the OODS staging area before ingestion. On "direct"
+        # ingestion, this is where the file is located.  This is a check
+        # to be sure that happened.
+        name = self.strip_prefix(self.destFile, self.forwarderStagingDir)
+        staged_file = os.path.join(self.stagingDirectory, name)
+        self.assertTrue(os.path.exists(staged_file))
+
+        # this file should now not exist
+        self.assertFalse(os.path.exists(self.destFile))
+
+        # add one more task, whose sole purpose is to interrupt the others by
+        # throwing an exception. This is used to exit all tasks.
+        task_list.append(asyncio.create_task(self.interrupt_me()))
+
+        return staged_file, task_list
+
+    async def testTaggedFileTestCase(self):
+        exposure = "3019053000001"
+        file_to_ingest, task_list = await self.stage()
+
+        # add an extra task, which runs after ingestion
+        task_list.append(asyncio.create_task(self.associate_file(exposure)))
+        task_list.append(asyncio.create_task(self.check_file(file_to_ingest)))
+        task_list.append(asyncio.create_task(self.disassociate_file(exposure)))
+        task_list.append(asyncio.create_task(self.check_file(file_to_ingest, wait=50, exists=False)))
+
+        # kick off all the tasks, until one (the "interrupt_me" task)
+        # throws an exception
+        try:
+            await asyncio.gather(*task_list)
+        except Exception:
+            for task in task_list:
+                task.cancel()
+
+    async def check_file(self, filename, wait=25, exists=True):
```

Comment:
Similarly, having a docstring here describing the purpose of the function and its arguments would be nice.

Reply:
added